src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java

```diff
@@ -0,0 +1,328 @@
+            }
+        });
+
+        final SAMSequenceDictionary sequences;
+        try {
+            sequences = BucketUtils.isHadoopUrl(ref) ?
+                    makeSequenceDictionary(getPath(URI.create(ref))) : makeSequenceDictionary(new File(ref));
+        } catch (IOException e) {
+            throw new GATKException("Cannot make sequence dictionary", e);
+        }
+        final SAMFileHeader readsHeader = new SAMFileHeader();
+        readsHeader.setSortOrder(getHeaderForReads().getSortOrder());
+        readsHeader.setSequenceDictionary(sequences);
+
+        final SAMLineParser samLineParser = new SAMLineParser(new DefaultSAMRecordFactory(), ValidationStringency.SILENT, readsHeader, null, null);
+        Broadcast<SAMLineParser> samLineParserBroadcast = ctx.broadcast(samLineParser);
+
+        JavaRDD<GATKRead> reads = samLines.map(r -> new SAMRecordToGATKReadAdapter(samLineParserBroadcast.getValue().parseLine(r)));
+
+        try {
+            ReadsSparkSink.writeReads(ctx, output, null, reads, readsHeader, ReadsWriteFormat.SHARDED);
+        } catch (IOException e) {
+            throw new GATKException("Unable to write bam",e);
+        }
+
+        reportThreadTimings();
+        log.info("Total time to run tool: " + toolStopwatch.elapsed(TimeUnit.SECONDS) + "s");
+    }
+
+    private <U> Iterator<List<U>> pairwise(Iterator<U> iterator) {
+        return Iterators.partition(iterator, 2);
+    }
+
+    private File localizeReferenceAndBwaIndexFiles() throws IOException {
+        Stopwatch downloadRefStopwatch = Stopwatch.createStarted();
+        File localRef = File.createTempFile("ref", ".fa");
+        localRef.delete();
+        Path localPath = localRef.toPath();
+        Path remotePath = getPath(URI.create(ref));
+        Files.copy(remotePath, localPath);
+        for (String extension : new String[] { ".amb", ".ann", ".bwt", ".pac", ".sa" }) {
+            Files.copy(remotePath.resolveSibling(remotePath.getFileName() + extension),
+                    localPath.resolveSibling(localPath.getFileName() + extension));
+        }
+        downloadRefStopwatch.stop();
+        log.info("Time to download reference: " + downloadRefStopwatch.elapsed(TimeUnit.SECONDS) + "s");
+        return localRef;
+    }
+
+    private static Path getPath(URI uri) throws IOException {
+        try {
+            return Paths.get(uri);
+        } catch (FileSystemNotFoundException e) {
+            // Try to use thread context classloader to find the filesystem. This can happen when the filesystem
+            // provider is loaded using a URL classloader (e.g. in spark-submit).
+            ClassLoader cl = Thread.currentThread().getContextClassLoader();
+            if (cl == null) {
+                throw e;
+            }
+            return FileSystems.newFileSystem(uri, new HashMap<>(), cl).provider().getPath(uri);
+        }
+    }
+
+    static <T> Iterator<T> concat(Iterator<? extends Iterable<T>> iterator) {
```

Comment:
can you add a comment what this method does? Also, make it private or add tests for it